File: Habitica/src/test/java/com/habitrpg/android/habitica/helpers/NumberAbbreviatorTest.java
package com.habitrpg.android.habitica.helpers;

import android.content.Context;

import com.habitrpg.android.habitica.R;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class NumberAbbreviatorTest {

    private Context context;

    @Before
    public void setUp() {
        context = mock(Context.class);

        when(context.getString(R.string.thousand_abbrev)).thenReturn("k");
        when(context.getString(R.string.million_abbrev)).thenReturn("m");
        when(context.getString(R.string.billion_abbrev)).thenReturn("b");
        when(context.getString(R.string.trillion_abbrev)).thenReturn("t");
    }

    @Test
    public void testThatItDoesntAbbreviatesSmallNumbers() {
        assertEquals("215", NumberAbbreviator.INSTANCE.abbreviate(context, 215));
        assertEquals("2.05", NumberAbbreviator.INSTANCE.abbreviate(context, 2.05));
    }

    @Test
    public void testThatItAbbreviatesThousand() {
        assertEquals("1.55k", NumberAbbreviator.INSTANCE.abbreviate(context, 1550));
    }

    @Test
    public void testThatItAbbreviatesMillion() {
        assertEquals("9.99m", NumberAbbreviator.INSTANCE.abbreviate(context, 9990000));
    }

    @Test
    public void testThatItAbbreviatesBillion() {
        assertEquals("1.99b", NumberAbbreviator.INSTANCE.abbreviate(context, 1990000000));
    }

    @Test
    public void testThatItAbbreviatesThousandWithoutAdditionalDecimals() {
        assertEquals("1k", NumberAbbreviator.INSTANCE.abbreviate(context, 1000));
        assertEquals("1.5k", NumberAbbreviator.INSTANCE.abbreviate(context, 1500));
    }

    @Test
    public void voidtestThatitRoundsCorrectly() {
        assertEquals("9.99k", NumberAbbreviator.INSTANCE.abbreviate(context, 9999));
    }

}

File: Habitica/src/test/java/com/habitrpg/android/habitica/helpers/NumberAbbreviatorTest.kt
package com.habitrpg.android.habitica.helpers

import android.content.Context
import com.habitrpg.android.habitica.R
import com.habitrpg.common.habitica.helpers.NumberAbbreviator.abbreviate
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.mockk
import java.util.Locale

class NumberAbbreviatorTest : StringSpec({
    val mockContext = mockk<Context>()
    beforeEach {
        Locale.setDefault(Locale.US)
        every { mockContext.getString(R.string.thousand_abbrev) } returns "k"
        every { mockContext.getString(R.string.million_abbrev) } returns "m"
        every { mockContext.getString(R.string.billion_abbrev) } returns "b"
        every { mockContext.getString(R.string.trillion_abbrev) } returns "t"
    }

    "should not abbreviate small numbers" {
        abbreviate(mockContext, 215.0, 2) shouldBe "215"
        abbreviate(mockContext, 2.05, 2) shouldBe "2.05"
    }

    "should abbreviate thousands" {
        abbreviate(mockContext, 1550.0, 2) shouldBe "1.55k"
    }

    "should abbreviate millions" {
        abbreviate(mockContext, 9990000.0, 2) shouldBe "9.99m"
    }

    "should abbreviate billions" {
        abbreviate(mockContext, 1990000000.0, 2) shouldBe "1.99b"
    }

    "should abbreviate trillions" {
        abbreviate(mockContext, 1990000000000.0, 2) shouldBe "1.99t"
    }

    "should abbreviate thousands without additional decimals" {
        abbreviate(mockContext, 1000.0, 2) shouldBe "1k"
        abbreviate(mockContext, 1500.0, 2) shouldBe "1.5k"
        abbreviate(mockContext, 1500.0, 0) shouldBe "1k"
    }

    "should round correctly" {
        abbreviate(mockContext, 9999.0, 2) shouldBe "9.99k"
    }

    "should force decimals for numbers between -1 and 1" {
        abbreviate(mockContext, 0.5, 0) shouldBe "0.5"
        abbreviate(mockContext, 0.3248, 0) shouldBe "0.32"
        abbreviate(mockContext, -0.99, 0) shouldBe "-0.99"
    }

    afterEach { clearMocks(mockContext) }
})
